chronos/bipbuffer.nim

```diff
@@ -0,0 +1,147 @@
+#
+#                     Chronos
+#
+#  (c) Copyright 2018-Present Status Research & Development GmbH
+#
+#                Licensed under either of
+#    Apache License, version 2.0, (LICENSE-APACHEv2)
+#                MIT license (LICENSE-MIT)
+
+## This module implements Bip Buffer (bi-partite circular buffer) by Simone
+## Cooke.
+##
+## The Bip-Buffer is like a circular buffer, but slightly different. Instead of
+## keeping one head and tail pointer to the data in the buffer, it maintains two
+## revolving regions, allowing for fast data access without having to worry
+## about wrapping at the end of the buffer. Buffer allocations are always
+## maintained as contiguous blocks, allowing the buffer to be used in a highly
+## efficient manner with API calls, and also reducing the amount of copying
+## which needs to be performed to put data into the buffer. Finally, a two-phase
+## allocation system allows the user to pessimistically reserve an area of
+## buffer space, and then trim back the buffer to commit to only the space which
+## was used.
+##
+## https://www.codeproject.com/Articles/3479/The-Bip-Buffer-The-Circular-Buffer-with-a-Twist
+
+{.push raises: [].}
+
+import results
+
+type
+  BipPos = object
+    start: Natural
+    finish: Natural
+
+  BipBuffer* = object
+    a, b, r: BipPos
+    data: seq[byte]
+
+proc init*(t: typedesc[BipBuffer], size: int): BipBuffer =
+  ## Creates new Bip Buffer with size `size`.
+  BipBuffer(data: newSeq[byte](size))
+
+template len(pos: BipPos): Natural =
+  pos.finish - pos.start
+
+template reset(pos: var BipPos) =
+  pos = BipPos()
+
+func init(t: typedesc[BipPos], start, finish: Natural): BipPos =
+  BipPos(start: start, finish: finish)
+
+func calcReserve(bp: BipBuffer): tuple[space: Natural, start: Natural] =
+  if len(bp.b) > 0:
+    (Natural(bp.a.start - bp.b.finish), bp.b.finish)
+  else:
+    let spaceAfterA = Natural(len(bp.data) - bp.a.finish)
+    if spaceAfterA >= bp.a.start:
+      (spaceAfterA, bp.a.finish)
+    else:
+      (bp.a.start, Natural(0))
+
+func availSpace*(bp: BipBuffer): Natural =
+  ## Returns amount of space available for reserve in buffer `bp`.
+  let (res, _) = bp.calcReserve()
+  res
+
+func len*(bp: BipBuffer): Natural =
+  ## Returns amount of used space in buffer `bp`.
+  len(bp.b) + len(bp.a)
+
+proc reserve*(bp: var BipBuffer, pt, st: typedesc,
+              size: Natural = 0): Result[tuple[data: pt, size: st], cstring] =
```

Comment:
```suggestion
              size: Natural = 0): tuple[data: pt, size: st] =
```

same here - I think we can make it a defect to try to reserve more than `bp.availSpace` - either you call this function with 0 or you can check for the available space with `availSpace` but I can't think of any situation where you call it with a larger value, that isn't a bug.

Reply:
the problem is this part here: `cast[st](len(bp.r))` - if I do `st=int8` this can easily return a non-sensical value - if I do `pt = int` it _should_ be divided by the size of `pt[]` but then you also need to consider alignment and padding, and the buffer might no longer be contiguous.. - `ptr byte` is the only "safe" thing to use here without opening up a pandora's box of problems

Reply:
Fixed in https://github.com/status-im/nim-chronos/pull/485/commits/ef7d4563e791733aeab7e0edc906e75773ceaf42